### PR TITLE
[RFC] vim-patch: 8.0.0041, 8.0.0042, 8.0.0043

### PR DESCRIPTION
**Pull request by tweekmonster**
vim-patch:8.0.0041
https://github.com/vim/vim/commit/869e35270ecffd9024958880cb03f6f0bb01ea93

```
Problem:    When using Insert mode completion but not actually inserting
            anything an undo item is still created. (Tommy Allen)
Solution:   Do not call stop_arrow() when not inserting anything.
```

vim-patch:8.0.0042
https://github.com/vim/vim/commit/cbd3bd6cbed5baf418b037b17ad46e339ff59174

```
Problem:    When using Insert mode completion with 'completeopt' containing
            "noinsert" change is not saved for undo.  (Tommy Allen)
Solution:   Call stop_arrow() before inserting for pressing Enter.
```

vim-patch:8.0.0043
https://github.com/vim/vim/commit/9ec7fa82a2c4f0283862ed704c2940959e6130ee

```
Problem:    When using Insert mode completion with 'completeopt' containing
            "noinsert" with CTRL-N the change is not saved for undo.  (Tommy
            Allen)
Solution:   Call stop_arrow() before inserting for any key.
```

### CODE REVIEW

**Comment by tweekmonster** [comment]
I'll reword and squash once the last issue is resolved: https://groups.google.com/d/msg/vim_dev/t66A6NIzLlc/7f0y7rdQBQAJ

File: src/nvim/edit.c [hunk 1]
```diff
@@ -3947,16 +3950,20 @@ static int ins_compl_get_exp(pos_T *ini)
...
+  // In replace mode: Put the old characters back, if any.
+  col = compl_col + (compl_cont_status & CONT_ADDING ? compl_length : 0);
+  if ((int)curwin->w_cursor.col > col) {
+    if (stop_arrow() == FAIL) {
+      return;
+    }
+    backspace_until_column(col);
+  }
+
+  // TODO(unknown): is this sufficient for redrawing?  Redrawing everything
```
Comment (tweekmonster) [inline src/nvim/edit.c]: The linter complained that there wasn't an assigned user for the `TODO`.  Is assigning it to `unknown` alright?

Comment (jamessan) [inline src/nvim/edit.c]: Did the TODO come from upstream?  If so, it looks like the precedent so far is `TODO(vim)`.

**Comment by tweekmonster** [comment]
This seems to have completely resolved the `complete()` issue in the past few days that I've used it.  Is there anything else that needs to be done before this is merged?

**Comment by justinmk** [comment]
Does the test fail correctly if the patches are not applied?

**Comment by tweekmonster** [comment]
Well, shit.  I just tried this I and I can't re-create the root issue (`complete()` overwriting undos) in nvim.  But, it was definitely there when I wrote the tests and they did in fact fail, but not any more.  I feel like I'm taking crazy pills.

But, I did forget to tweak the test to deal with superfluous entries being created.  I'll do that now.

**Comment by tweekmonster** [comment]
Tests on master will fail now:

```
Failure → test/functional/viml/completion_spec.lua @ 338
completion completeopt+=noinsert does not add blank undo items using <c-n>, <c-p>, <c-y>
test/functional/viml/completion_spec.lua:363: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'foo'
  [2] = 'bar'
 *[3] = 'foo' }
Expected:
(table) {
  [1] = 'foo'
  [2] = 'bar'
 *[3] = 'ccc' }

stack traceback:
        test/functional/viml/completion_spec.lua:363: in function <test/functional/viml/completion_spec.lua:338>

Failure → test/functional/viml/completion_spec.lua @ 338
completion completeopt+=noinsert does not add blank undo items using <up>, <down>, <cr>
test/functional/viml/completion_spec.lua:363: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'foo'
  [2] = 'bar'
 *[3] = 'foo' }
Expected:
(table) {
  [1] = 'foo'
  [2] = 'bar'
 *[3] = 'ccc' }

stack traceback:
        test/functional/viml/completion_spec.lua:363: in function <test/functional/viml/completion_spec.lua:338>
```

**Comment by justinmk** [comment]
Thank you @tweekmonster !

**Comment by tweekmonster** [comment]
@justinmk No, thank you for putting an end to the unreasonable amount of attention I've given this!  I take it that the CI failures are unrelated?

**Comment by justinmk** [comment]
CI failures were: 1 flaky UBSAN test; 1 gcov nonsense; 1 appveyor being slow.

And thanks for giving attention to that, I also am very sensitive to empty undo steps. `gww` and `gqq` are going to drive me crazy some day (or drive me to spend days finally fixing them)

**Pull request merged**